### DISCUSSION TITLE
[RFC] `turbo run` (without tasks) and `turbo ls`

### DISCUSSION THREAD

**Discussion by anthonyshew** [vercel/turborepo#8811] [RFC]
In Turborepo 2.0.9, we've introduced experimental subcommands for repository discoverability. This is especially important in large monorepos, where many teams work together and it's useful to gain a quick understanding of the [Package and Task Graphs](https://turbo.build/repo/docs/core-concepts/package-and-task-graph).

- `turbo run`: Print a list of the tasks available to be used with `turbo run`.
  - This is the existing `turbo run` command, but with no tasks listed.
- `turbo ls`: Print a list of the packages in the repository.
  - Without arguments: List all packages in the repository.
  - `turbo ls <package>`: Print the dependencies of a specific package.
  
Both of these commands work with [`--filter`](https://turbo.build/repo/docs/reference/run#--filter-string) to refine the printed output to your intended subset of your workspace.
  
To help us make sure these tools are providing you with useful information, we appreciate you trying out these commands on your repository and providing your feedback.

### In a repository using `turbo@>2.0.9`

Run the commands above.

### In a repository using `turbo@<2.0.9`

You can avoid upgrading your repository by using `npx` for a one-off `turbo` command.
```
npx turbo@latest run --skip-infer
npx turbo@latest ls --skip-infer
npx turbo@latest ls <package> --skip-infer
```

We look forward to hearing your feedback on this feature!

**Comment by Willem-Jaap** [comment]
Nice!

I ran this in my repo, the list of commands becomes very long and maybe can be compacted down a bit 
![image](https://github.com/user-attachments/assets/fe6aba30-ded7-49da-896c-ac0cbd3bd1ae)


Maybe a column layout?
| Task       | Packages                                                                      |
| ---------- | ----------------------------------------------------------------------------- |
| lint       |  @spark/admin, @spark/api, @spark/frontend, @spark/frontend-e2e, @spark/types |
| dev        | @spark/admin, @spark/api, @spark/frontend                                     |
| dev:turbo  |  @spark/admin, @spark/api, @spark/frontend                                    |

**Comment by Netail** [comment]
The current `turbo ls` output is quite overwhelming/cluttered (can maybe even be a little confusing to read through), maybe something like;

![image](https://github.com/user-attachments/assets/5fd6a0f8-9fb1-4461-997f-c30c4cbb3064)

vs the old

![image](https://github.com/user-attachments/assets/ff5197f7-4857-4f0a-92cf-d23816aa0ab6)

**Comment by anthonyshew** [comment]
This feature has been implemented! Thank you to all who gave feedback on this RFC.

If you have issues to report or feature requests related to this feature, please create an Issue or Discussion accordingly!
